Update section: When to use Async Completion (#2078)
temporalio/documentation: docs-src/concepts/what-is-asynchronous-activity-completion.md

#### When to use Async Completion

The intended use case for this feature is when an external system has the final result of a computation, started by an Activity.
However, often it works better for the Activity to notify the external system, complete normally, and then wait in the Workflow for a [Signal](/concepts/what-is-a-signal) from the external system.

> Consider using Async Completion when

Author: @djmagee I think no colon is what you said to do when I want to avoid "in the following scenarios:"?

Author: (which feels wordy to me)

Reviewer: Right, per https://developers.google.com/style/colons#introductory-phrase-preceding-colon.

> - the external process is unreliable and might fail to send critical status updates through a Signal
> - you want the external process to Heartbeat or receive Cancellation

Reviewer: ```suggestion
- the external process is unreliable and might fail to send critical status updates through a Signal.
- you want the external process to Heartbeat or receive Cancellation.
```
Per https://developers.google.com/style/lists#numbered-lettered-bulleted-lists.

> Consider using Signals when you want to immediatly retry a failure to notify the external system.

Reviewer 2: ```suggestion
Consider using Signals when you want to immediately retry a failure to notify the external system.
```

> Consider using Signals when you want to immediatly retry a failure to notify the external system.

Reviewer 2: I think this is not worded the best (we encourage retry policy to govern retries and you can set to immediate retry). I believe what @mfateev is saying in the internal thread is "Consider using Signals when you only want the activity success/fail/retry/timeouts to govern the _start_ of the operation rather than the whole operation". I'm not sure how best to word that.

Author: > we encourage retry policy to govern retries and you can set to immediate retry

this reads to you as not using a retry policy? perhaps:

Consider using Signals when you want to retry sooner a failure to notify the external system.

I think it's helpful to talk about the desired result here, and then get into details below

Reviewer 2: Unsure how to word it best. I was just confused by the original statement which at first glance felt like it was saying "if you want to immediately retry failure, don't use async activity" which I don't think is what we wanted to say.

> Immediate retries are helpful when the external process might take long time.

Reviewer: ```suggestion
Immediate retries are helpful when the external process might take a long time.
```
Or could be "...might be lengthy."

> For example, if the external process is waiting for a human to review something and respond, and may take days, then if you set a [Start-To-Close Timeout](/activities#start-to-close-timeout) of one week on an Activity with Async Completion, it's possible that the Activity fails to notify the external system and doesn't throw an error (for example, if the Worker dies).

Reviewer: ```suggestion
For example, consider an external process that is waiting for a human to review something and respond, and might take days.
If you set a [Start-To-Close Timeout](/activities#start-to-close-timeout) of one week on an Activity with Async Completion, the Activity might fail to notify the external system and not throw an error (for example, if the Worker dies).
```

> In this case, the Activity will be retried in a week.

Reviewer: ```suggestion
In this case, the Activity is retried in a week.
```

It would be better to be able to retry the Activity immediately.
If the Activity completes normally after notifying the external system, it can have a short Start-To-Close Timeout of 10 seconds (for example, if the external system is notified by an API call that should never take more than 10 seconds).

> In this case, if the notification fails, it will be retried after 10 seconds.

Reviewer: ```suggestion
In this case, if the notification fails, the Activity is retried after 10 seconds.
```

#### When to use Async Completion

The intended use case for this feature is when an external system has the final result of a computation, started by an Activity.
However, often it works better for the Activity to notify the external system, complete normally, and then wait in the Workflow for a [Signal](/concepts/what-is-a-signal) from the external system.

Consider using Async Completion when

- the external process is unreliable and might fail to send critical status updates through a Signal.
- you want the external process to Heartbeat or receive Cancellation.

> Consider using Signals when you want to retry sooner a failure to notify the external system.

Reviewer 3: I am not sure what this sentence is trying to say tbh.
Maybe a missing word? Sooner than what?

Immediate retries are helpful when the external process might take a long time.
For example, consider an external process that is waiting for a human to review something and respond, and might take days.

> If you set a [Start-To-Close Timeout](/activities#start-to-close-timeout) of one week on an Activity with Async Completion, the Activity might fail to notify the external system and not throw an error (for example, if the Worker dies).

Reviewer 3: Notify the external system of what?
I think this example depends on a mapped out information flow -
ATM I find it hard to keep track of that model in my head when I read it.

Would you be willing to completely lay out the scenario from the perspective of both so that it is easier to compare?

Also - it sounds like Async completion has a very limited use case - Is there one we can call out explicitly here?
